File: tests/golden/orm/jpa/postgresql/relations.kt
/*
 * grabado が生成した Jakarta Persistence の entity（Kotlin）。
 *
 * **kotlin("plugin.jpa") が要る** —— JPA は引数の無いコンストラクタを求めるので、
 * コンストラクタ引数で書く形はプラグイン前提になる（Spring Initializr の既定）。
 * package 宣言は出さない（置き場所は生成物を受け取る側が決める）。
 */

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.IdClass
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import java.io.Serializable

@Entity
@Table(name = "employees")
class Employee(
    @Id
    @Column(name = "id")
    var id: Int,

    @Column(name = "name", nullable = false)
    var name: String,

    /** 直属の上長（自己参照） */
    @ManyToOne
    @JoinColumn(name = "manager_id", nullable = true)
    var manager: Employee? = null,
)

@Entity
@Table(name = "projects")
class Project(
    @Id
    @Column(name = "id")
    var id: Int,

    @Column(name = "title", nullable = false)
    var title: String,

    @ManyToOne
    @JoinColumn(name = "owner_id", nullable = false)
    var owner: Employee,

    @ManyToOne
    @JoinColumn(name = "team_id", nullable = true)
    var team: Team? = null,
)

@Entity
@Table(name = "teams")
class Team(
    @Id
    @Column(name = "id")
    var id: Int,

    @Column(name = "name", nullable = false)
    var name: String,
)

@Entity
@Table(name = "employee_projects")
@IdClass(EmployeeProjectId::class)
class EmployeeProject(
    @Id
    @Column(name = "employee_id")
    var employeeId: Int,

    @Id
    @Column(name = "project_id")
    var projectId: Int,
)

/** employee_projects の複合主キー（JPA は @IdClass に id クラスを要求する） */
data class EmployeeProjectId(
    var employeeId: Int? = null,
    var projectId: Int? = null,
) : Serializable

File: tests/golden/orm/drizzle/postgresql/relations.ts
// grabado が生成した Drizzle のスキーマ。
//
// **型は core ごとに違う** —— pg / mysql / sqlite で関数名も表せる意味も
// 変わるので、正規型からの表を core ごとに持っている（段階6-9f）。
//
// **TypeScript の識別子は ASCII だけ**なので、非 ASCII の名前は `_` に潰して通し番号で
// 一意化してある。**元の名前は型関数の第 1 引数に必ず残る**。

import type { AnyPgColumn } from "drizzle-orm/pg-core";
import { integer, pgTable, primaryKey, text } from "drizzle-orm/pg-core";

export const employees = pgTable("employees", {
    id: integer("id").primaryKey(),
    name: text("name").notNull(),
    /** 直属の上長（自己参照） */
    managerId: integer("manager_id").references((): AnyPgColumn => employees.id),
});

export const projects = pgTable("projects", {
    id: integer("id").primaryKey(),
    title: text("title").notNull(),
    ownerId: integer("owner_id").notNull().references(() => employees.id),
    teamId: integer("team_id").references(() => teams.id),
});

export const teams = pgTable("teams", {
    id: integer("id").primaryKey(),
    name: text("name").notNull(),
});

export const employeeProjects = pgTable("employee_projects", {
    employeeId: integer("employee_id").notNull().references(() => employees.id),
    projectId: integer("project_id").notNull().references(() => projects.id),
}, (t) => [
    primaryKey({ columns: [t.employeeId, t.projectId] }),
]);